### DISCUSSION TITLE
Renaming Doctrine Annotation alias

### DISCUSSION THREAD

**Discussion by etshy** [rectorphp/rector#8195] [Q&A] [answered]
Hi everyone
I discovered Rector recently and tried to make a Rule to rename the alias on my Doctrine entities.

On the projects I work on, Previous team member decided to not follow the convention and didn't aliased the Doctrine Mapping folder (or used another alias that `ORM`), so I have things like this :
```php

use Doctrine\ORM\Mapping;

class MyClass
{
    /**
    * @Mapping\column
    */
    $property;
}
```

And it seems the Doctrine set to change annotation to attributes only work with ORM alias (at least the @Mapping annotation doesn't seem to be changed in the dry-run detected changes)

So I wanted to make a rule to an alias to my use (if not already existing) and rename `@Mapping` to `@ORM\` but I can't seem to do it.

I tried making a rule for properties only (just to test) and looping through the `phpDocInfo->children`, changing the name and returning the node if something changed in the node.

Something like this (I don't have the code I made right now as It's on the work pc)
```php

foreach($phpDocNode->children as $key $child) {
    ...
    $child->name = str_replace('@Mapping', '@ORM', $child->name);
}
...
return $node;
...
```

I also tried unsetting the `child` for the `children` array and re-adding the `child` in the array (in case it cause some "markedAsChanged" process?) but didn't work.

And of course i can't use the `RenameClassRector` because of the alias on Folder.

How could I rename the aliases using Rector ?

ps : As I said I can't show you the code I did right now, so I hope it's clear enough.

**Comment by etshy** [comment]
I managed to do what I wanted with the following code (lots of hardcoded things as it's still in testing) : 
```php
class RenmeMappingAnnotationRector extends AbstractRector
{
    public function __construct(
        private readonly UseImportsResolver $useImportsResolver,
        private readonly DocBlockUpdater $docBlockUpdater,
    ) {
    }

    public function getRuleDefinition(): RuleDefinition
    {
        return new RuleDefinition("", []);
    }

    public function getNodeTypes(): array
    {
        return [Node\Stmt\Property::class, Node\Stmt\Class_::class];
    }

    public function refactor(Node $node)
    {
        $phpDocInfo = $this->phpDocInfoFactory->createFromNode($node);

        if ($phpDocInfo === null) {
            return null;
        }

        $this->checkUseImportsAndImportORMIfNeeded($phpDocInfo);

        $hasDocBlockChanged = $this->replaceAnnotation(
            $phpDocInfo,
            "@Mapping\\",
            "@ORM\\"
        );

        if ($hasDocBlockChanged) {
            $this->docBlockUpdater->updateRefactoredNodeWithPhpDocInfo($node);
            $hasChanged = true;
        }

        if (!$hasChanged) {
            return null;
        }

        return $node;
    }

    private function checkUseImportsAndImportORMIfNeeded(PhpDocInfo $phpDocInfo): void
    {
        $uses = $this->useImportsResolver->resolveBareUses();
        $aliasAlreadyImported = false;

        foreach ($uses as $keyUse => $use) {
            if (empty($use->uses)) {
                continue;
            }
            foreach ($use->uses as $keyUseUse => $useUse) {
                if ($useUse->name->toString() !== "Doctrine\\ORM\\Mapping") {
                    continue;
                }
                if ($useUse->alias !== null && $useUse->alias->toString() === "ORM") {
                    //ORM alias already exists
                    $aliasAlreadyImported = true;
                    break 2;
                }
            }
        }


        foreach ($uses as $keyUse => $use) {
            if (empty($use->uses)) {
                continue;
            }

            $useUseHasChanged = false;
            $useUseRemoved = false;

            foreach ($use->uses as $keyUseUse => $useUse) {
                if ($useUse->name->toString() !== "Doctrine\\ORM\\Mapping") {
                    continue;
                }
                if ($useUse->alias !== null && $useUse->alias->toString() === "ORM") {
                    //ORM alias already exists
                    continue;
                }

                //found Mapping namespace not aliased import
                //we remove it
                unset($use->uses[$keyUseUse]);


                if ($aliasAlreadyImported) {
                    $useUseRemoved = true;
                    continue;
                }
                $useUseHasChanged = true;
                //if not already imported we add the alias
                $useUse->alias = new Node\Identifier("ORM");
                $use->uses[$keyUseUse] = $useUse;

                ksort($use->uses);
                $use->uses = array_values($use->uses);
            }

            if ($useUseHasChanged || $useUseRemoved) {
                if ($use->uses === []) {
                    unset($uses[$keyUse]);
                } else {
                    $uses[$keyUse] = $use;
                }
                ksort($uses);
                $uses = array_values($uses);
            }
        }
    }

    private function replaceAnnotation(PhpDocInfo $phpDocInfo, string $oldTag, string $newTag): bool
    {
        $hasChanged = false;
        $phpDocNode = $phpDocInfo->getPhpDocNode();
        foreach ($phpDocNode->children as $key => $phpDocChildNode) {
            if (!$phpDocChildNode instanceof PhpDocTagNode) {
                continue;
            }

            if (!$phpDocChildNode->value instanceof DoctrineAnnotationTagValueNode) {
                continue;
            }

            if (!str_contains($phpDocChildNode->name, $oldTag)) {
                continue;
            }

            unset($phpDocNode->children[$key]);
            $newValue = $phpDocChildNode->value;
            $newValue->identifierTypeNode = new IdentifierTypeNode(
                str_replace("@Mapping\\", "@ORM\\", $newValue->identifierTypeNode->name)
            );
            $phpDocNode->children[$key] = new SpacelessPhpDocTagNode(
                str_replace("@Mapping\\", "@ORM\\", $phpDocChildNode->name),
                $newValue
            );
            //sort by key to avoid moving annotation too much
            ksort($phpDocNode->children);
            $phpDocNode->children = array_values($phpDocNode->children);

            $hasChanged = true;
        }

        return $hasChanged;
    }
}
```

But I have a problem the removal of use statement. With my code in `checkUseImportsAndImportORMIfNeeded` I end with empty use statements : 
```php
-use Doctrine\ORM\Mapping;
+use ;
```

I checked the `UseImportsRemover` but it does the same I did, unset `Use_` and `UseUse`. And i tried using it I end up with the same problem.

Is there any known fix to remove use statements ?

**Comment by samsonasik** [comment, marked as answer]
Hi, removing use statements should on top level node `FileWithoutNamespace` or `Namespace`.

First, include `FileWithoutNamespace` and `Namespace` on after target node on `getNodeTypes()` method:

https://github.com/laminas/laminas-servicemanager-migration/blob/cdedb23f9df8480daead4dbef79b35f26fedcc86/src/Rector/Class_/ImplementsFactoryInterfaceToPsrFactoryRector.php#L64

then, verify after it can be removed:

https://github.com/laminas/laminas-servicemanager-migration/blob/1.42.x/src/Rector/Class_/ImplementsFactoryInterfaceToPsrFactoryRector.php#L72-L73

**Reply by etshy** [reply]
It worked !

I hesitated to add Namespace in the target, because I thought I could remove as easily as I could edit them thanks to the resolver, I should have tried it before.

I guess `FileWithoutNamespace` si targeting php file without namespace ? I don't htink I'll need it for this vere specific Rule I'm making.

Anyways, thanks a lot for the help !

**Reply by samsonasik** [reply]
Yes, if you have php file that doesn't have namespace, you can add `FileWithoutNamespace`

**Comment by etshy** [comment]
Seems like it doesn't work that much...

On nested Annotation it does weird things
```php
- *
- * @Mapping\Table(
- *     name="entity",
- *     uniqueConstraints={@UniqueConstraint(name="access_unique", columns={"column1", "column2"})}
- * )
- * @Mapping\Entity(repositoryClass="path\to\Repository")
- * @Mapping\HasLifecycleCallbacks()
- */
+ *
+ * @ORM\Table(name="entity", uniqueConstraints={(name="access_unique", columns={"column1", "column2"})})
+ * @ORM\Entity(repositoryClass="path\to\Repository")
+ * @ORM\HasLifecycleCallbacks()
+ */
```

I guess it's because I only create a new `identifierTypeNode` without recreating the nested values ?
Though I don't understand how I can recreate these values

**Reply by etshy** [reply]
I tried update my method like this 
```php
private function replaceAnnotation(PhpDocInfo $phpDocInfo, string $oldTag, string $newTag): bool
    {
        $hasChanged = false;
        $phpDocNode = $phpDocInfo->getPhpDocNode();
        foreach ($phpDocNode->children as $key => $phpDocChildNode) {
            if (!$phpDocChildNode instanceof SpacelessPhpDocTagNode) {
                continue;
            }

            if (!$phpDocChildNode->value instanceof DoctrineAnnotationTagValueNode) {
                continue;
            }

            if (!str_contains($phpDocChildNode->name, $oldTag)
                || !str_contains($phpDocChildNode->value->identifierTypeNode->name, $oldTag)
            ) {
                continue;
            }

            //unset($phpDocNode->children[$key]);

            $phpDocChildNode->name = str_replace($oldTag, $newTag, $phpDocChildNode->name);
            $phpDocChildNode->value->identifierTypeNode->name = str_replace($oldTag, $newTag, $phpDocChildNode->value->identifierTypeNode->name);
            $phpDocChildNode->setAttribute('start_and_end', null);

            $phpDocNode->children[$key] = $phpDocChildNode;

            //sort by key to avoid moving annotation too much
            ksort($phpDocNode->children);
            $phpDocNode->children = array_values($phpDocNode->children);
            $hasChanged = true;
        }

        return $hasChanged;
    }
```

I'm not making any new object, I just change what I want and set `start_and_end` to null to trigger a change.
I though that would keep the `parent` relation between each level, but I still have `@UniqueContraint` that disappear.
And I have no idea why and how.

**Reply by etshy** [reply]
I also noticed that array are acting weird, event though I didn't edited them.
For example
```php
-     *
-     * @Mapping\Column(name="locked", type="boolean", options={"default": false}, nullable=false)
-     */
+     *
+     * @Mapping\Column(name="locked", type="boolean", options={"default"=false}, nullable=false)
+     */
```

the `:`  was changed to `=` for no real reason.

Maybe that's related to the Annotation disappearing from the above message as both are related to array ?

**Reply by samsonasik** [reply]
as this is your custom rector rules, you may need to debug and find the node types :)

**Reply by etshy** [reply]
I used dd to dump the node after my edition. 
`@ORM\Table` still had a value with `@UniqueConstraint` as name (and I don't modify nested annotations yet in my code so this "level" shouldn't change at all) and the `options` array for `@Column` have no reason to change, as i only change the name of the DoctrineAnnotation node and the Identifier.

So for now I don't know what cause the annotation to disappear and the array to change

i'm a bit lost with the PHPParser, that's the first I'm working with and it's really complex to understand

**Reply by etshy** [reply]
I tracked down the print issue (for the `@UniqueConstraint` annotation) and I found it comes for the method `Rector\BetterPhpDocParser\PhpDoc\DoctrineAnnotationTagValueNode::__toString`.

The method seems to only care about the content of the node without checking if it's a nested annotation.

I changed the method from
```php
    public function __toString() : string
    {
        if (!$this->hasChanged) {
            if ($this->originalContent === null) {
                return '';
            }
            return $this->originalContent;
        }
        if ($this->values === []) {
            if ($this->originalContent === '()') {
                // empty brackets
                return $this->originalContent;
            }
            return '';
        }
        $itemContents = $this->printValuesContent($this->values);
        return \sprintf('(%s)', $itemContents);
    }
```
to 
```php
    public function __toString() : string
    {
        if (!$this->hasChanged) {
            if ($this->originalContent === null) {
                return '';
            }
            return $this->originalContent;
        }
        if ($this->values === []) {
            if ($this->originalContent === '()') {
                // empty brackets
                return $this->originalContent;
            }
            return '';
        }

        $doctrineIdentifier = '';
        if (
            $this->identifierTypeNode instanceof IdentifierTypeNode
            //on first level the parent is a SpacelessPhpDocTagNode and the toString method alreay add the name
            && !$this->getAttribute('parent') instanceof SpacelessPhpDocTagNode
        ) {
            $doctrineIdentifier .= (string)$this->identifierTypeNode;
        }

        $itemContents = $this->printValuesContent($this->values);
        return \sprintf('%s(%s)', $doctrineIdentifier, $itemContents);
    }
```

It works for my case but I haven't tested further, so it very probably will break other things.

**Reply by samsonasik** [reply]
Feel free to create feature request PR :)

**Reply by etshy** [reply]
Will do in my free time

**Reply by etshy** [reply]
Hey @samsonasik I tried installing `rector-src` to work on the PR, but I have a few problems (I'm on windows btw)

Following the contributing page on rector site, the command `composer fix-cs` throw me access denied error on my php Temp folder and is modifying every file to CRLF. Is there a way to keep LF on windows ?

`composer phpstan` is working great but report 11 errors

and a lot of unit Tests are failing.

Should I install WSL2 and work on WSL2 instead of windows ?

**Reply by samsonasik** [reply]
I don't have windows on my local dev so I am not sure if I can help, feel free to provide a patch when you can :)

**Reply by etshy** [reply]
I'll try to take a look to fix rector-src on windows first then.
Don't expect anything soon though :)

**Reply by etshy** [reply]
Well there's too much failing tests coming from Windows file system, I will have to rewrite, or skip for Windows, too much Unit tests.
I'll just install a WSL2 Debian and work on the PR from there.

**Comment by etshy** [comment]
Sorry to add more to this discussion !

I had to update my `replaceAnnotation` method to this
```php
private function replaceAnnotation(PhpDocInfo $phpDocInfo, string $oldTag, string $newTag): bool
    {
        $hasChanged = false;
        $phpDocNode = $phpDocInfo->getPhpDocNode();
        foreach ($phpDocNode->children as $key => $phpDocChildNode) {

            $phpDocChildNode->setAttribute('start_and_end',null);

            if (!$phpDocChildNode instanceof SpacelessPhpDocTagNode) {
                continue;
            }

            if (!$phpDocChildNode->value instanceof DoctrineAnnotationTagValueNode) {
                continue;
            }

            if (!str_contains($phpDocChildNode->name, $oldTag)
                || !str_contains($phpDocChildNode->value->identifierTypeNode->name, $oldTag)
            ) {
                continue;
            }

            $phpDocChildNode->name = str_replace($oldTag, $newTag, $phpDocChildNode->name);
            $phpDocChildNode->value->identifierTypeNode->name = str_replace(
                $oldTag,
                $newTag,
                $phpDocChildNode->value->identifierTypeNode->name
            );

            //trigger change
            $phpDocChildNode->setAttribute('start_and_end',null);

            $phpDocNode->children[$key] = $phpDocChildNode;
            //sort by key to avoid moving annotation too much
            ksort($phpDocNode->children);
            $phpDocNode->children = array_values($phpDocNode->children);
            $hasChanged = true;
        }

        return $hasChanged;
    }
```

I had to put the `setAttribute('start_and_end',null)` for every node for the whole dockblock to re-render. (didn't find another way to fore the render other than putting `start_and_end` to null).

Without that, if the annotation I want to rename are NOT at the end, the old one will comme back 
via this code
```php
     if ($startAndEnd instanceof StartAndEnd && !$shouldReprintChildNode) {
            $isLastToken = $nodeCount === $key;
            // correct previously changed node
            $this->correctPreviouslyReprintedFirstNode($key, $startAndEnd);
            $output = $this->addTokensFromTo($output, $this->currentTokenPosition, $startAndEnd->getEnd(), $isLastToken);
            if (str_contains($output, '@Mapping')) {
                //dd($this->currentTokenPosition, $startAndEnd->getEnd(), $isLastToken);
            }
            $this->currentTokenPosition = $startAndEnd->getEnd();
            return \rtrim($output);
        }
```
in `PhpDocInfoPrinter::printDocChildNode`

Here is the diff I had before my change in `trplaceAnnotation`
```php
 /**
  * Class CommandExport
  * @package App\Domain\Tenant\Export
- *
+ *
+ * @ORM\Table(name="table")
+ * @ORM\Entity(repositoryClass="tableRepo")
  * @Mapping\Table(name="table")
  * @Mapping\Entity(repositoryClass="tableRepo")
  *
  * @ref test
- */
+ */
```
As you can see, the old node changed from `@Mapping` to `@ORM` then the `@Mapping` were recovered during the re-printing (I'm using `DocBlockUpdater::updateRefactoredNodeWithPhpDocInfo($node)` in my rule)

Is there a better way to trigger a change than changing the StartEnd attribute ? 
Without having to recreate the whole PhpDocNode, because I really don't know how to do that ^^".
